### PR TITLE
New Substring, Left & Right Wrappers

### PR DESCRIPTION
**Pull request by 304NotModified**
Examples:
```c#
${substring:${level}:start=2:length=2}    // from index 2, length 2
${substring:${level}:start=-2:length=2}   // from end -2, length 2
${left:${level}:length=2}    // from start, length 2
${right:${level}:length=2}   // from end, length 2
```

- [x] negative start -> measure for end
- [x] negative length? -> empty result
- [x] use WrapperLayoutRendererBase
- [x] review notes

### CODE REVIEW

File: src/NLog/LayoutRenderers/Wrappers/LeftLayoutRendererWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            substringWrapper = new SubstringLayoutRendererWrapper();
+        }
+
+        /// <summary>
+        /// Gets or sets the length in characters. 
+        /// </summary>
+        /// <value>Index</value>
+        /// <docgen category='Transformation Options' order='10' />
+        [RequiredParameter]
+        public int Length
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/Wrappers/LeftLayoutRendererWrapper.cs]: default 1?

**Comment by snakefoot** [comment]
Instead of doing a StringBuilder.Remove, then you should consider to allocate resulting-string, and truncate StringBuilder (assigning Length to zero), and then append string afterwards (Ofcouse if should just truncate from the right, then just truncate the StringBuilder by assigning Length to wanted value).

File: src/NLog/LayoutRenderers/Wrappers/LeftLayoutRendererWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+namespace NLog.LayoutRenderers.Wrappers
+{
+    /// <summary>
+    /// Left part of a text
+    /// </summary>
+    [LayoutRenderer("left")]
+    [ThreadAgnostic]
+    public class LeftLayoutRendererWrapper : WrapperLayoutRendererBuilderBase
+    {
+        private SubstringLayoutRendererWrapper substringWrapper;
```
Comment (NLog-SonarQube) [inline src/NLog/LayoutRenderers/Wrappers/LeftLayoutRendererWrapper.cs]: MAJOR Make 'substringWrapper' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

File: src/NLog/LayoutRenderers/Wrappers/RightLayoutRendererWrapper.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+namespace NLog.LayoutRenderers.Wrappers
+{
+    /// <summary>
+    /// Right part of a text
+    /// </summary>
+    [LayoutRenderer("right")]
+    [ThreadAgnostic]
+    public class RightLayoutRendererWrapper : WrapperLayoutRendererBuilderBase
+    {
+        private SubstringLayoutRendererWrapper substringWrapper;
```
Comment (NLog-SonarQube) [inline src/NLog/LayoutRenderers/Wrappers/RightLayoutRendererWrapper.cs]: MAJOR Make 'substringWrapper' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

**Comment by NLog-SonarQube** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by 304NotModified** [comment]
> Instead of doing a StringBuilder.Remove, then you should consider to allocate resulting-string, and truncate StringBuilder (assigning Length to zero), and then append string afterwards (Ofcouse if should just truncate from the right, then just truncate the StringBuilder by assigning Length to wanted value without allocating string).

@snakefoot you meant this? -> https://github.com/NLog/NLog/pull/2194/commits/e56835864544ecae06cbc70601d189be15ea5cbf

**Comment by snakefoot** [comment]
You should not inherit from BuilderBase but just override RenderInnerAndTransform
________________________________

**Comment by snakefoot** [comment]
Yes. Trying to get rid of the other :)



Sent from my Samsung device


-------- Original message --------

**Comment by 304NotModified** [comment]
done

although I don't like 

https://github.com/NLog/NLog/blob/f51c86b33f020cb566181a3a3c20e437d840aeb6/src/NLog/LayoutRenderers/Wrappers/LeftLayoutRendererWrapper.cs#L71

**Comment by 304NotModified** [comment]
I think this is ready. @snakefoot please check if you like

**Comment by snakefoot** [comment]
Believe you can override initialize method where you can do such assignment (and also initialize the nested layout renderer)

Btw think you can optimize the substring layoutrenderer so it just truncated the StringBuilder without allocating a substring. Ex for doing Left-Truncate

Sent from my Samsung device


-------- Original message --------

**Comment by snakefoot** [comment]
Your DoSubstring method need to recognize that the StringBuilder provided can already be in use. It should only perform substring operation on the extra output generated from the Inner-Layout. That is why the orgLength parameter is there to help recognize where transform operation should start from.



Sent from my Samsung device


-------- Original message --------

**Comment by 304NotModified** [comment]
> Believe you can override initialize method where you can do such assignment (and also initialize the nested layout renderer)

That is a problem if Inner will be changed after initialize?



> Your DoSubstring method need to recognize that the StringBuilder provided can already be in use. I

I added a check on the result of renderInner (https://github.com/NLog/NLog/pull/2194/commits/392eb724c7a4abab12f6509a4341fb62f998bcb3)

I think that is sufficient. If now, please let me know. 

I like to merge this one soon

**Comment by snakefoot** [comment]
You still don't understand. Try read my previous comment again. Your wrapper only works if only one layoutrenderer is used in the layout. Try put some fixed text in front and see if it remains untouched. You need to use orgLength


Sent from my Samsung device


-------- Original message --------

**Comment by snakefoot** [comment]
Yes orgLength of the StringBuilder is the start position for rendering of the inner-layout into the StringBuilder and the start position for doing the transformation afterwards



Sent from my Samsung device


-------- Original message --------

**Comment by 304NotModified** [comment]
Ah my 'Clear(..)` was wrong.

Added tests and fixed the issue

**Comment by 304NotModified** [comment]
Added also a note. It's a bit tricky design.

**Comment by 304NotModified** [comment]
do you mean

```
   builder.Append("myString", 1, 10);
```

?

**Comment by 304NotModified** [comment]
updated.

It would be nice to defer rendering the inner until fully needed (or partial render), but I don't think that's possible now.

**Comment by snakefoot** [comment]
@304NotModified Now created #2907. Maybe you understand code better than sentences.

**Comment by 304NotModified** [comment]
superseded by #2907

**Pull request closed without merging**